### PR TITLE
In skin cmp fix aus

### PR DESCRIPTION
**Pull request by ioanna0**
## What does this change?

Co-authored-by: @mxdvl 

Depend on `willShowPrivacyMessage` to decide if an inskin should be shown. All the rest of the changes are for converting the `getPageTargeting` calls to asynchronous nature

## Does this change need to be reproduced in dotcom-rendering ?

- [x] No
- [ ] Yes (please indicate your plans for DCR Implementation)

### Tested

- [ ] Locally
- [ ] On CODE (optional)

### CODE REVIEW

**Comment by mxdvl** [comment]
This looks like the right approach!

**Comment by mxdvl** [comment]
Temporary fix: https://github.com/guardian/frontend/pull/23281

**Comment by mxdvl** [comment]
There are still conflicts to resolve, but all test are passing on TeamCity.

I've also simplified some tests in ea517c1.

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:306]: this is wrong now, I think it should be:

```suggestion
const getPageTargeting = (): Promise<{ [key: string]: mixed }> => {
```

and then because `getPageTargeting` is exported, everything that uses it will need to be updated?

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:307]: this is wrong now, I think it should be:

```suggestion
const getPageTargeting = (): Promise<{ [key: string]: mixed }> => {
    if (Object.keys(myPageTargetting).length !== 0) return Promise.resolve(myPageTargetting);
```

and then because getPageTargeting is exported, everything that uses it will need to be updated?

**Comment by mxdvl** [comment]
Lots of missing types, for a subsequent PR:

```ts
type PageTargettingMissing = {
    ab: string[],
    af: string,
    at: string,
    br: string,
    bl: string[],
    cc: string,
    cmp_interaction: string,
    consent_tcfv2: string,
    fr: string,
    ms: string,
    ob: string,
    pa: string,
    ref: string,
    rdp: string,
    rp: string,
    se: string[],
    si: string,
    vl: string,
+
```

File: static/src/javascripts/projects/commercial/modules/header-bidding/prebid/prebid.js
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/modules/header-bidding/prebid/prebid.js:121]: This `async` constructor is a little different than the original one. To use the object, we now need to wait for `PrebidAdUnit.build()` to resolve.

File: static/src/javascripts/projects/commercial/modules/header-bidding/types.js
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/modules/header-bidding/types.js:101]: Enforcing `bidParams` to always return a Promise seemed the cleanest approach, but there are still many instances where it could remain synchronous.

Having a mix of types, as below, seemed worse and harder to reason about:

```
        | Promise<PrebidSonobiParams>
        | PrebidIndexExchangeParams
        | PrebidTrustXParams
        | Promise<PrebidOpenXParams>
        | PrebidOzoneParams
        | Promise<PrebidAdYouLikeParams>
        | PrebidPubmaticParams
```

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:48]: We extracted this `type` as it's being used in various places in the code.

In a further PR, we should review the `PageTargetting` type and make it handle all cases.

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:310]: If this function is `async`, this could simply return the page targetting object, as `async` functions always return a Promise.

```suggestion
        return myPageTargetting;
```

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.spec.js
Comment (mxdvl) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.spec.js:354]: Sorry, these simplifications are for a subsequent PR.

**Review by mxdvl — COMMENTED** [review]
Added some comments to guide the review.

File: static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js
Comment (sndrs) [inline static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js:61]: probably not - I bet this just kicked of `onConsentChange` - def worth checking after this is merged

File: static/src/javascripts/projects/commercial/modules/header-bidding/types.js
Comment (sndrs) [reply static/src/javascripts/projects/commercial/modules/header-bidding/types.js:101]: yeah def - if it needs to be async in one instance, all should be imo

File: static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js
Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js:61]: That's @ioanna and I's thought exactly.

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:308]: I think this function can be simplified:

```js
const getPageTargeting = (): Promise<PageTargettingLoose> =>
    new Promise(resolve => {
        if (Object.keys(myPageTargetting).length !== 0)
            return resolve(myPageTargetting);

        onConsentChange(async state => {
            let canRun: boolean | null;
            if (state.ccpa) {
                // CCPA mode
                canRun = !state.ccpa.doNotSell;
            } else if (state.tcfv2) {
                // TCFv2 mode
                canRun = state.tcfv2.consents
                    ? Object.keys(state.tcfv2.consents).length > 0 &&
                      Object.values(state.tcfv2.consents).every(Boolean)
                    : false;
            } else if (state.aus) {
                // AUS mode
                canRun = state.aus.personalisedAdvertising;
            } else canRun = false;

            if (canRun !== latestConsentCanRun) {
                const ccpaState = state.ccpa ? state.ccpa.doNotSell : null;
                const eventStatus = state.tcfv2
                    ? state.tcfv2.eventStatus
                    : 'na';
                myPageTargetting = await buildPageTargetting(
                    canRun,
                    ccpaState,
                    eventStatus
                );
                latestConsentCanRun = canRun;
                resolve(myPageTargetting);
            }
        });
    });
```

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:311]: because this is an async function, it automatically returns a promise. therefore you don't need to wrap `pageTargeting` here

Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:311]: It's not wrapped anymore…

**Review by sndrs — APPROVED** [review]
a couple of comments but no blocker – looks good to me, and not too frightening broken down 👍 

really good effort @ioanna0 and @mxdvl

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [reply static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:311]: sorry, that's on the wrong line! deleting and re-adding

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:305]: because this is an async function, it automatically returns a promise. therefore you don't need to wrap `pageTargeting` here

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (mxdvl) [reply static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:308]: I’m thinking we could introduce `getConsentFor('framework', state)` or `getConsentBoolean(state)`. 

Looking at how other consumers of the CMP are using it, the granularity of TCFv2 is rarely leveraged.

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (ashishpuliyel) [inline static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js:340]: This may work, but it feels really wierd to me. This is a callback that's fired in three ways:
1 - We're waiting for CMP to initalize, it'll be fired when it's done.
2 - CMP already initalized, fired immediately with current state.
3 - Consent has changed (?) So every time someone goes into the privacy manager?

In case 3, where it's already fired once, what's the meaning of 'resolve'. You can only resolve a promise once and it was resolved when the CMP initalised the first time.

Discuss on chat when you see this.

File: static/src/javascripts/projects/commercial/modules/header-bidding/types.js
Comment (ashishpuliyel) [reply static/src/javascripts/projects/commercial/modules/header-bidding/types.js:101]: Agree. The only other option I can think of is bidParams and bidParamsAsync, and the consumer checks both (or only one if it knows which it needs)

**Review by ashishpuliyel — CHANGES_REQUESTED** [review]
Lets discuss getPageTargeting

File: static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js
Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js]: I’ve added a test for this, and we are now `await`ing: 7550dfe

Comment (sndrs) [reply static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js]: did flow pick that up?

File: static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js
Comment (sndrs) [reply static/src/javascripts/projects/common/modules/commercial/build-page-targeting.js]: will `buildPageTargetting` return different results if consent state changes? if so then yeah resolving once here isn't enough

File: static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js
Comment (ashishpuliyel) [reply static/src/javascripts/projects/commercial/modules/header-bidding/prebid/bid-config.js]: Apparently not.

**Comment by mxdvl** [comment]
The changes are a little too wide-ranging, and we’re working on a better proposal for how the page targetting should be built and how to handle consent changes.

**Comment by mxdvl** [comment]
We’ve decided to carry on with the change, but pass a state to `getPageTargetting`

**Comment by mxdvl** [comment]
After further considerations, changing the signature of `getPageTargetting` to accept the `state` as an argument is not workable.

An alternative approach leveraging Promises and their resolved state can be found in https://github.com/guardian/frontend/pull/23330

**Pull request closed without merging**
